### ISSUE TITLE
 IWD2: Uncanny Dodge breaks save game compatibility

### ISSUE THREAD

**Issue by MarcelHB** [gemrb/gemrb#2160]
#### Bug description
If your thief reaches level 3, there is the _Uncanny Dodge_ feature obtained. Apparently we have encoded this as effect no. 459 to be applied.

Problem: IWD2 does not like unknown effect codes.

#### Screenshots
![iwd2_unknown_opcode](https://github.com/user-attachments/assets/8c28d3e9-303a-49cb-b797-91f27efec7c3)

GemRB version (check as many as you know apply)
- [x] master as of this issue
- [ ] 0.9.3
- [ ] 0.9.2
- [ ] 0.9.1

**Comment by lynxlynxlynx** [comment]
Yeah, same for Spell focus, turn undead3 and maybe more.
https://gibberlings3.github.io/iesdp/opcodes/iwd2.htm#op457

**Comment by MarcelHB** [comment]
Also SpellFocus (398) gets attached to NPCs and persists in the save game.

**Comment by lynxlynxlynx** [comment]
This now hopefully won't happen any more, but a resave in gemrb is needed first.

**Comment by MarcelHB** [comment]
I resaved one such a game and it still pops up, same message.

**Comment by lynxlynxlynx** [comment]
If you ctrl-m on the actors to get their effect queues, what's the highest opcode number?

**Comment by MarcelHB** [comment]
0x1CB

**Comment by lynxlynxlynx** [comment]
Ok and I guess there's no effect name printed. I forgot that the saved queues just store the opcode numbers, so our association does not matter. We'd need a translator in EFFImporter or CREImporter to sanitize these interim saves.

**Comment by lynxlynxlynx** [comment]
I hope this helps, resaving should move them into an empty range. Spellfocus has already been there, so it shouldn't trigger an assert nor have any effect in the original.

**Comment by MarcelHB** [comment]
It now complains about effect id 392 being unknown.

Maybe I can find a trace back to the original check, just in case it's not just an upper bound but also disallows the gaps.

**Comment by lynxlynxlynx** [comment]
Bleh, it's the default case in its large switch case it uses for decoding. We'll need to rethink this.

**Comment by MarcelHB** [comment]
Where is it even encoded that a thief receives Uncanny Dodge at level 3? Unlike _Dodge_ it's not a feat to choose. I see some extra skill/feat stuff in the LU script, but I do not find this example.

**Comment by lynxlynxlynx** [comment]
It's in CLABTH01.2da, shipped with the game. We then parse that mess and apply `gemrb/unhardcoded/iwd2/39280.spl`. 

I've made some good progress and cleaned up IESDP in this regard. Found a few opcode slots we can safely reuse, but I'll reevaluate if all are needed.

**Comment by MarcelHB** [comment]
A strref to set this up? Oh dear, a mess indeed.

**Comment by lynxlynxlynx** [comment]
You can retry now. The interim save you made is busted though, something from before yesterday's changes is needed for testing.

**Comment by MarcelHB** [comment]
Looks good from what I tested.

With some later savegames, I now run into a new issue much earlier in loading but that's then a different story, I presume.

**Comment by lynxlynxlynx** [comment]
So saves from new games (you can cheat in some xp) don't hit this any more, right?

**Comment by MarcelHB** [comment]
Yes, looks good, can use them in the original now. From the savegame in #2432 yet I see something in GemRB that looks suspicous wrt this issue:

```
[Python/ERROR]:   File "D:/Sources\gemrb\gemrb\GUIScripts/LUCommon.py", line 478, in ApplyFeats
    IDLUCommon.LearnFeatInnates (MyChar, MyChar < 1000, False) # did we get passed a global id?
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[Python/ERROR]:   File "D:/Sources\gemrb\gemrb\GUIScripts\iwd2/IDLUCommon.py", line 223, in LearnFeatInnates
    SetSpellFocus (pc)
    ~~~~~~~~~~~~~~^^^^
[Python/ERROR]:   File "D:/Sources\gemrb\gemrb\GUIScripts\iwd2/IDLUCommon.py", line 201, in SetSpellFocus
    GemRB.ApplyEffect (pc, "SpellFocus", bonus, i, "", "", "", "SPLFOCUS", 1)
    ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[Python/ERROR]: RuntimeError: Invalid effect name!
```

**Comment by lynxlynxlynx** [comment]
Ah, that's simple.
